src/main/java/fact/io/hdureader/Fits.java

```diff
@@ -0,0 +1,194 @@
+
+            }
+
+        } catch (EOFException e){
+            primaryHDU = hdus.get(0);
+            stream.close();
+
+            log.debug("A total of {} HDUs were found in the file.", hdus.size());
+        }
+    }
+
+    static DataInputStream getDecompressedDataStream(URL url) throws IOException {
+        InputStream stream = url.openStream();
+
+        byte[] header = new byte[2];
+        stream.mark(2);
+
+        int read = stream.read(header);
+        if(read != 2){
+            throw new IOException("Could not read the first 2 bytes from stream.");
+        }
+
+        //reset stream to start
+        stream.reset();
+
+        //check if matches standard gzip magic number
+        if(isGzippedCompressed(header)) {
+            log.debug("Getting gzipped stream");
+            return new DataInputStream(new BufferedInputStream(new GZIPInputStream(stream)));
+        }else {
+            return new DataInputStream(new BufferedInputStream(stream));
+        }
+    }
+
+    /**
+     * Given the first 2 bytes of an input stream, this returns whether the stream is
+     * gzipped or not
+     * @param header byte[] containing the first two bytes of any file/stream
+     * @return true iff stream is gzipped.
+     */
+    public static boolean isGzippedCompressed(byte[] header){
+        return header[0] == (byte) GZIPInputStream.GZIP_MAGIC  && header[1] == (byte) (GZIPInputStream.GZIP_MAGIC >> 8);
+    }
+
+    /**
+     * Each extension has a keyword called EXTNAME. This method returns the extension with the given name if it exists.
+     *
+     * @param extname the HDU to get.
+     * @return the HDU with the passed EXTNAME value
+     */
+    public HDU getHDU(String extname){
+        return hduNames.get(extname);
+    }
+
+
+    public Optional<BinTable> getBinTableByName(String hduExtName) {
+        if(!hduNames.containsKey(hduExtName)){
+            return Optional.empty();
+        }
+        HDU hdu = hduNames.get(hduExtName);
+        return Optional.ofNullable(hdu.getBinTable());
+    }
+
+
+    /**
+     * Provides a datastream to the data area of the given hdu.
+     * This method is useful when reading custom data extensions that are present in the fits file.
+     * @param hdu the HDU of interest.
+     * @return a DataInputStream providing data from the data section of the given hdu.
+     * @throws IOException in case the stream cannot be opened
+     */
+    public DataInputStream getInputStreamForHDUData(HDU hdu) throws IOException {
+        DataInputStream stream = getDecompressedDataStream(url);
+
+
+        long bytesToSkip = 0;
+        for(HDU h : hdus){
+            if(h.equals(hdu)){
+                bytesToSkip += h.header.headerSizeInBytes;
+                break;
+            }
+            bytesToSkip += h.header.headerSizeInBytes + h.sizeOfDataArea();
+        }
+
+
+        long skipped = stream.skip(bytesToSkip);
+        if(bytesToSkip!= skipped ){
```

Comment:
It might be yes. I'm a bit reluctant to wrap the stream in another layer. Not sure if that actually improves readability. I could add another functions which performs this check. But you know. Then its another method with a side effect.